specs/gloas/beacon-chain.md

```diff
@@ -1048,6 +1146,209 @@ def process_operations(state: BeaconState, body: BeaconBlockBody) -> None:
+        return
+    # Verify the validator has been active long enough
+    if get_current_epoch(state) < validator.activation_epoch + SHARD_COMMITTEE_PERIOD:
+        return
+
+    pending_balance_to_withdraw = get_pending_balance_to_withdraw(state, validator_index)
+    if withdrawal_request.amount == FULL_EXIT_REQUEST_AMOUNT:
+        # Only exit validator if it has no pending withdrawals in the queue
+        if pending_balance_to_withdraw == 0:
+            initiate_validator_exit(state, validator_index)
+        return
+
+    has_sufficient_effective_balance = validator.effective_balance >= MIN_ACTIVATION_BALANCE
+    has_excess_balance = (
+        state.balances[validator_index] > MIN_ACTIVATION_BALANCE + pending_balance_to_withdraw
+    )
+
+    # Only allow partial withdrawals with compounding withdrawal credentials
+    if (
+        has_compounding_withdrawal_credential(validator)
+        and has_sufficient_effective_balance
+        and has_excess_balance
+    ):
+        to_withdraw = min(
+            state.balances[validator_index] - MIN_ACTIVATION_BALANCE - pending_balance_to_withdraw,
+            withdrawal_request.amount,
+        )
+        exit_queue_epoch = compute_exit_epoch_and_update_churn(state, to_withdraw)
+        withdrawable_epoch = Epoch(exit_queue_epoch + MIN_VALIDATOR_WITHDRAWABILITY_DELAY)
+        state.pending_partial_withdrawals.append(
+            PendingPartialWithdrawal(
+                validator_index=validator_index,
+                amount=to_withdraw,
+                withdrawable_epoch=withdrawable_epoch,
+            )
+        )
+```
+
+###### Modified `process_withdrawal_request`
+
+*Note*:
+
+```python
+def process_withdrawal_request(state: BeaconState, withdrawal_request: WithdrawalRequest) -> None:
+    # If partial withdrawal queue is full, only full exits are processed
+    is_queue_full = len(state.pending_partial_withdrawals) == PENDING_PARTIAL_WITHDRAWALS_LIMIT
+    is_full_exit_request = withdrawal_request.amount == FULL_EXIT_REQUEST_AMOUNT
+    if is_queue_full and not is_full_exit_request:
+        return
+
+    builder_pubkeys = [b.pubkey for b in state.builders]
+    if withdrawal_request.validator_pubkey in builder_pubkeys:
+        index = builder_pubkeys.index(withdrawal_request.validator_pubkey)
+        builder_index = BuilderIndex(index)
+        return process_withdrawal_request_for_builder(state, withdrawal_request, builder_index)
+
+    validator_pubkeys = [v.pubkey for v in state.validators]
+    if withdrawal_request.validator_pubkey in validator_pubkeys:
+        index = validator_pubkeys.index(withdrawal_request.validator_pubkey)
+        validator_index = ValidatorIndex(index)
+        return process_withdrawal_request_for_validator(state, withdrawal_request, validator_index)
+```
+
+##### Deposit requests
+
+###### New `get_index_for_new_builder`
+
+```python
+def get_index_for_new_builder(state: BeaconState) -> BuilderIndex:
```

Comment:
Flagging this function as a possible problem. Don't have it clear yet but it's something that clients may want to discuss in the breakout. We typically have a global pubkey -> index (or inverse) cache for signature validation and this type of reusage would make this cache a little more sensitive to be in sync with the head view. 

Not against this, but just flagging as a possible shot in our feet.